File: packages/asset-compiler-core/topology.mjs
function edgeKey(a,b){return a<b?`${a},${b}`:`${b},${a}`;}
function addLinkEdge(adj,a,b){
 if(a===b)return;
 if(!adj.has(a))adj.set(a,new Set());
 if(!adj.has(b))adj.set(b,new Set());
 adj.get(a).add(b);adj.get(b).add(a);
}
function componentCount(adj){
 const seen=new Set();let count=0;
 for(const start of adj.keys()){
  if(seen.has(start))continue;
  count++;
  const stack=[start];
  while(stack.length){
   const node=stack.pop();
   if(seen.has(node))continue;
   seen.add(node);
   for(const next of adj.get(node)??[])if(!seen.has(next))stack.push(next);
  }
 }
 return count;
}
function classifyVertexLink(adj){
 if(adj.size===0)return 'locked';
 let degreeOne=0,degreeTwo=0;
 for(const neighbours of adj.values()){
  const degree=neighbours.size;
  if(degree===1)degreeOne++;
  else if(degree===2)degreeTwo++;
  else return 'locked';
 }
 if(componentCount(adj)!==1)return 'locked';
 if(degreeOne===0&&degreeTwo===adj.size)return 'interior';
 if(degreeOne===2&&degreeTwo===adj.size-2)return 'boundary';
 return 'locked';
}
/** Half-edge classification. Does not repair topology. Vertex count is the POSITION accessor length. */
export function classifyTopology(indices,vertexCount){
 if(!Number.isSafeInteger(vertexCount)||vertexCount<0)throw new Error('vertexCount is required');
 if(!Array.isArray(indices)&&!(indices instanceof Uint32Array)&&!(indices instanceof Uint16Array))throw new Error('indices are required');
 if(indices.length%3!==0)throw new Error('Index count must be a positive multiple of three');
 const edges=new Map();
 const links=Array.from({length:vertexCount},()=>new Map());
 const incident=new Uint32Array(vertexCount);
 for(let face=0,offset=0;offset<indices.length;face++,offset+=3){
  const tri=[Number(indices[offset]),Number(indices[offset+1]),Number(indices[offset+2])];
  for(const vertex of tri)if(!Number.isInteger(vertex)||vertex<0||vertex>=vertexCount)throw new Error('Invalid index');
  for(let e=0;e<3;e++){
   const start=tri[e],end=tri[(e+1)%3];
   const key=edgeKey(start,end);
   if(!edges.has(key))edges.set(key,[]);
   edges.get(key).push({start,end,face});
   incident[start]++;
   addLinkEdge(links[start],end,tri[(e+2)%3]);
  }
 }
 let boundary=0,manifold=0,nonManifold=0;
 const neighbors=Array.from({length:indices.length/3},()=>new Set());
 for(const occurrences of edges.values()){
  if(occurrences.length===1)boundary++;
  else if(occurrences.length===2&&occurrences[0].start===occurrences[1].end&&occurrences[0].end===occurrences[1].start){
   manifold++;
   neighbors[occurrences[0].face].add(occurrences[1].face);
   neighbors[occurrences[1].face].add(occurrences[0].face);
  }else nonManifold++;
 }
 let interior=0,boundaryVertices=0,locked=0,unused=0;
 for(let vertex=0;vertex<vertexCount;vertex++){
  if(incident[vertex]===0){unused++;continue;}
  const kind=classifyVertexLink(links[vertex]);
  if(kind==='interior')interior++;
  else if(kind==='boundary')boundaryVertices++;
  else locked++;
 }
 return {
  triangles:indices.length/3,
  edges:{boundary,manifold,nonManifold},
  vertices:{interior,boundary:boundaryVertices,locked,unused},
  neighbors:neighbors.map(set=>[...set].sort((a,b)=>a-b)),
  manifold:nonManifold===0&&locked===0,
 };
}

File: packages/asset-compiler-rust/src/topology.rs
pub(crate) mod link;
use crate::{invalid, Result};
use link::{classify_link, LinkScratch};
#[derive(Debug, Clone, PartialEq)]
pub struct TopologyReport {
    pub triangles: usize,
    pub boundary_edges: usize,
    pub manifold_edges: usize,
    pub non_manifold_edges: usize,
    pub interior_vertices: usize,
    pub boundary_vertices: usize,
    pub locked_vertices: usize,
    pub unused_vertices: usize,
    pub manifold: bool,
}
fn edge_key(a: u32, b: u32) -> (u32, u32) {
    if a < b {
        (a, b)
    } else {
        (b, a)
    }
}
/// Edge and vertex classification of a triangle soup. The cluster DAG builds its own cluster
/// graph, so no triangle adjacency is derived here.
pub fn classify_topology(indices: &[u32], vertex_count: usize) -> Result<TopologyReport> {
    if !indices.len().is_multiple_of(3) || indices.is_empty() {
        return Err(invalid("Index count must be a positive multiple of three"));
    }
    let triangle_count = indices.len() / 3;
    let mut halves = Vec::with_capacity(triangle_count * 3);
    let mut incident = vec![0u32; vertex_count];
    let mut link_count = vec![0u32; vertex_count];
    for face in 0..triangle_count {
        let tri = [
            indices[face * 3],
            indices[face * 3 + 1],
            indices[face * 3 + 2],
        ];
        for vertex in tri {
            if vertex as usize >= vertex_count {
                return Err(invalid("Invalid index"));
            }
        }
        for e in 0..3 {
            let start = tri[e];
            let end = tri[(e + 1) % 3];
            let key = edge_key(start, end);
            halves.push((key.0, key.1, start, end, face));
            incident[start as usize] += 1;
            if start != end {
                link_count[start as usize] += 1;
            }
        }
    }
    // Compressed links: one flat array with a per-vertex offset, instead of a vector per vertex.
    let mut offsets = vec![0u32; vertex_count + 1];
    for vertex in 0..vertex_count {
        offsets[vertex + 1] = offsets[vertex] + link_count[vertex];
    }
    let mut cursor = offsets.clone();
    let mut links = vec![(0u32, 0u32); offsets[vertex_count] as usize];
    for face in 0..triangle_count {
        let tri = [
            indices[face * 3],
            indices[face * 3 + 1],
            indices[face * 3 + 2],
        ];
        for e in 0..3 {
            let start = tri[e];
            let end = tri[(e + 1) % 3];
            let other = tri[(e + 2) % 3];
            if start == end {
                continue;
            }
            let at = &mut cursor[start as usize];
            links[*at as usize] = (end, other);
            *at += 1;
        }
    }
    halves.sort_unstable_by(|left, right| {
        left.0
            .cmp(&right.0)
            .then(left.1.cmp(&right.1))
            .then(left.4.cmp(&right.4))
    });
    let mut boundary = 0;
    let mut manifold = 0;
    let mut non_manifold = 0;
    let mut cursor = 0usize;
    while cursor < halves.len() {
        let mut end = cursor + 1;
        while end < halves.len()
            && halves[end].0 == halves[cursor].0
            && halves[end].1 == halves[cursor].1
        {
            end += 1;
        }
        let n = end - cursor;
        if n == 1 {
            boundary += 1;
        } else if n == 2
            && halves[cursor].2 == halves[cursor + 1].3
            && halves[cursor].3 == halves[cursor + 1].2
        {
            manifold += 1;
        } else {
            non_manifold += 1;
        }
        cursor = end;
    }
    let mut interior = 0;
    let mut boundary_vertices = 0;
    let mut locked = 0;
    let mut unused = 0;
    let mut scratch = LinkScratch::default();
    for vertex in 0..vertex_count {
        if incident[vertex] == 0 {
            unused += 1;
            continue;
        }
        let span = offsets[vertex] as usize..offsets[vertex + 1] as usize;
        match classify_link(&links[span], &mut scratch) {
            "interior" => interior += 1,
            "boundary" => boundary_vertices += 1,
            _ => locked += 1,
        }
    }
    Ok(TopologyReport {
        triangles: triangle_count,
        boundary_edges: boundary,
        manifold_edges: manifold,
        non_manifold_edges: non_manifold,
        interior_vertices: interior,
        boundary_vertices,
        locked_vertices: locked,
        unused_vertices: unused,
        manifold: non_manifold == 0 && locked == 0,
    })
}
#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn triangle_is_boundary_manifold() {
        let report = classify_topology(&[0, 1, 2], 3).expect("topology");
        assert_eq!(report.boundary_edges, 3);
        assert_eq!(report.manifold_edges, 0);
        assert_eq!(report.boundary_vertices, 3);
        assert!(report.manifold);
    }
    #[test]
    fn tetrahedron_is_closed_manifold() {
        let report = classify_topology(&[0, 1, 2, 0, 3, 1, 1, 3, 2, 0, 2, 3], 4).expect("topology");
        assert_eq!(report.boundary_edges, 0);
        assert_eq!(report.manifold_edges, 6);
        assert_eq!(report.interior_vertices, 4);
        assert!(report.manifold);
    }
    #[test]
    fn three_triangles_on_one_edge_are_non_manifold() {
        let report = classify_topology(&[0, 1, 2, 0, 1, 3, 0, 1, 4], 5).expect("topology");
        assert_eq!(report.non_manifold_edges, 1);
        assert!(!report.manifold);
        assert!(report.locked_vertices >= 2);
    }
}
